src/app/dataset-view/dataset.service.ts

```diff
@@ -303,66 +230,140 @@ export class AppDatasetService {
+                    };
+                }
+            ),
+            upstreamDependencies: originMetadata.currentUpstreamDependencies.map(
+                (upDependency) => {
+                    return {
+                        basics: upDependency as DatasetBasicsFragment,
+                        downstreamDependencies: [],
+                        upstreamDependencies: upDependency.metadata.currentUpstreamDependencies.map(
+                            (upDependency2) => {
+                                return {
+                                    basics: upDependency2 as DatasetBasicsFragment,
+                                    downstreamDependencies: [],
+                                    upstreamDependencies: upDependency2.metadata.currentUpstreamDependencies.map(
+                                        (upDependency3) => {
+                                            return {
+                                                basics: upDependency3 as DatasetBasicsFragment,
+                                                downstreamDependencies: [],
+                                                upstreamDependencies: upDependency3.metadata.currentUpstreamDependencies.map(
+                                                    (upDependency4) => {
+                                                        return {
+                                                            basics: upDependency4 as DatasetBasicsFragment,
+                                                            downstreamDependencies: [],
+                                                            upstreamDependencies: upDependency4.metadata.currentUpstreamDependencies.map(
+                                                                (upDependency5) => {
+                                                                    return {
+                                                                        basics: upDependency5 as DatasetBasicsFragment,
+                                                                        downstreamDependencies: [],
+                                                                        upstreamDependencies: []
+                                                                    };
+                                                                }
+                                                            )
+                                                        };
+                                                    }
+                                                )
+                                            };
+                                        }
+                                    )
+                                };
+                            }
+                        )
+                    };
+                }
+            )
+        };
+     }
+
+    private updatelineageGraph(origin: DatasetLineageNode) {
+        const lineageGraphEdges: DatasetBasicsFragment[][] = [];
```

Comment:
Should the dealings with nodes and edges be a concern of Lineage component, not dataset service? `origin` node seems to contain everything that is needed to vizualize the graph and the way it happens is not a service-level concern?

Reply:
At least building a traditional graph data structure should be done in the service on my opinion. The service should hide the complexity of not completely adequate GraphQL structure. But the bondary can be moved as we revise it again.